### PR TITLE
📋 Add complete flow tests for FTLA

### PR DESCRIPTION
**Pull request by skalermo**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
     "test:true-currencies": "mocha 'test/true-currencies/*.ts'",
     "test:true-gold": "mocha 'test/true-gold/*.ts'",
     "test:truefi": "mocha 'test/truefi/**/*.ts'",
-    "test:truefi2": "NODE_OPTIONS='--max-old-space-size=5120' mocha 'test/truefi2/**/*.ts' --exclude test/truefi2/lines-of-credit/**/*.ts test/truefi2/complete-flow/**/*.ts",
+    "test:truefi2": "NODE_OPTIONS='--max-old-space-size=5120' mocha 'test/truefi2/**/*.ts' --exclude test/truefi2/lines-of-credit/**/*.ts --exclude test/truefi2/complete-flow/**/*.ts",
```
Comment (MiksuJak) [inline package.json:28]: Was complete-flow running 2 times before that?

Comment (skalermo) [reply package.json:28]: Yep.

Comment (vanruch) [reply package.json:28]: 👍

File: test/truefi2/complete-flow/FixedTermLoans.test.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    // safu redeems the debt
+    await safu.redeem(debtToken.address)
+
+    // borrower stays banned even after full repayment of defaulted loan
+    expect(await borrowingMutex.locker(borrower.address)).to.eq('0x0000000000000000000000000000000000000001')
+
+    const loanInterest = await loan.interest()
+    expect(await tusd.balanceOf(pool.address)).to.eq(poolBalanceBefore.add(loanInterest))
+    expect(await tusd.balanceOf(safu.address)).to.eq(safuBalanceBefore)
+  })
```
Comment (vanruch) [inline test/truefi2/complete-flow/FixedTermLoans.test.ts:127]: Let's add a happy case when loan is repaid

**Review by MiksuJak — APPROVED** [review]

File: test/truefi2/complete-flow/FixedTermLoans.test.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    await tusd.connect(borrower).approve(loan.address, parseEth(5e5))
+    await loan.repay(borrower.address, parseEth(5e5))
+
+    // no further repayments, end of the term, after grace period, loan gets defaulted
+    await timeTravel(2 * MONTH)
+    await timeTravel((await creditOracle.gracePeriod()).toNumber())
+    tx = loan.enterDefault()
+    const debtToken = (await extractDebtTokens(loanFactory, owner, tx))[0]
+
+    // borrower pays part of unpaid debt to defaulted loan token
```
[the remark this answers is not in the document]
Comment (skalermo) [reply test/truefi2/complete-flow/FixedTermLoans.test.ts:102]: Do you think this file is a good place for it? I could just extract it into a separate test.

[the remark this answers is not in the document]
Comment (MiksuJak) [reply test/truefi2/complete-flow/FixedTermLoans.test.ts:102]: No, not necessarily this file. Let's create a new task for it.

[the remark this answers is not in the document]
Comment (skalermo) [reply test/truefi2/complete-flow/FixedTermLoans.test.ts:102]: Task filed.

File: test/truefi2/complete-flow/FixedTermLoans.test.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    // borrower repays the debt
+    const debt = await debtToken.totalSupply()
+    await tusd.connect(borrower).approve(debtToken.address, debt)
+    await debtToken.repayInFull(borrower.address)
+
+    // safu redeems the debt
+    await safu.redeem(debtToken.address)
+
+    // borrower stays banned even after full repayment of defaulted loan
+    expect(await borrowingMutex.locker(borrower.address)).to.eq('0x0000000000000000000000000000000000000001')
```
[the remark this answers is not in the document]
Comment (skalermo) [reply test/truefi2/complete-flow/FixedTermLoans.test.ts]: I replaced it only here leaving another instances to replace for another PR.

File: test/utils/constants.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import { constants } from 'ethers'
 
 export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
```
Comment (skalermo) [inline test/utils/constants.ts:3]: It seems in the past for zero address different constant was used. I'd prefer to stick with `AddressZero`. Let's remove this one.

**Review by MiksuJak — APPROVED** [review]

**Review by vanruch — APPROVED** [review]

**Pull request merged**